R/emisCO2.R

```diff
@@ -22,483 +22,746 @@
-  emis_cc <- .emis_nac(gdx, b = dummy, type = "cc")
-  emis_cc <- .emis_ac(gdx, b = emis_cc, type = "cc")
-  
-  ### emis_regrowth
-  # vegc,litc; soilc=0 #direct human + natural effects
-  # calc regrowth emissions by multiplication of age-class specific area with
-  # increase of carbon density in age-classes according to the time step length
-  # example for 2005; previous time step 2000 -> 5 year time step length
-  # emis regrowth 2005 = ((90 tc/ha in ac45 in 2005 - 100 tc/ha in ac50 in 2005)
-  #                                  * 10 Mha in 2005) / 5 years = -20 MtC/yr
-  emis_regrowth <- .emis_ac(gdx, b = dummy, type = "regrowth")
-  
-  ### emis_degrad
-  # calc emissios from forest degradation (secdforest and primforest),
-  # e.g. shifting_agriculture
-  # emis_degrad = (area damaged * carbon density) / time step length
-  emis_degrad <- dummy
-  p35_disturbance_loss_secdf <- readGDX(gdx, "p35_disturbance_loss_secdf", react = "quiet")
-  p35_disturbance_loss_primf <- readGDX(gdx, "p35_disturbance_loss_primf", react = "quiet")
-  if (all(!is.null(p35_disturbance_loss_secdf), !is.null(p35_disturbance_loss_primf))) {
-    agPools <- c("vegc", "litc")
-    pm_carbon_density_ac <- readGDX(gdx, "pm_carbon_density_ac")[, getYears(timestep_length), ]
-    emis_degrad[, , "secdforest"][, , agPools] <- dimSums(p35_disturbance_loss_secdf * pm_carbon_density_ac[, , agPools], dim = "ac")
-    fm_carbon_density <- readGDX(gdx, "fm_carbon_density")[, getYears(emis_degrad), ]
-    names(dimnames(fm_carbon_density))[2] <- "t"
-    emis_degrad[, , "primforest"][, , agPools] <- p35_disturbance_loss_primf * fm_carbon_density[, , "primforest"][, , agPools]
-  }
-  emis_degrad <- emis_degrad / timestep_length
-  emis_degrad[, 1, ] <- NA
-  
-  
-  ## Emissions from deforestation ---
-  # For land types with age classes, deforestation emissions are calculated as the difference between
-  # realized forest area and a counterfactual forest area, where all forest area in age class x is 
-  # assumed to progress to age class x+1 in time t+1.
-  .calcChangeAC <- function(forest) {
-    
-    counterfactual <- forest
-    counterfactual[ , , ] <- 0
-    
-    ageClasses <- getItems(forest, dim = 3)
-    years <- getItems(forest, dim = 2)
-    for (t in seq(from = 2, to = length(years))) {
-      for (ac in seq(from = 2, to = length(ageClasses))) {
-        if (ageClasses[ac] == "acx") {
-          # The largest age class simply accumulates
-          counterfactual[ , t, ac] <- forest[ , t - 1, "ac155"] + forest[ , t - 1, "acx"] 
-        } else {
-          counterfactual[ , t, ac] <- forest[ , t - 1, ac - 1]
-        }
-      }
+
+  ###
+  # various checks for output validation
+  .validateCalculation <- function(totalStock, totalStockCheck, output) {
+
+    # --- Ensure independent output of carbonstock is nearly equivalent to own calculation
+    if (any(totalStock - totalStockCheck > 1e-05, na.rm = TRUE)) {
```

Comment:
Kristine, this check had to be 1e-05. Florian suggested that the issue could be someplace here: https://github.com/magpiemodel/magpie/blob/master/modules/59_som/cellpool_jan23/presolve.gms. 

However, since it's still such a small inconsistency, it doesn't warrant too serious an investigation.